fontique/src/charmap.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright 2024 the Parley Authors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use read_fonts::{
+    FontData, FontRead, FontRef, TableProvider, TopLevelTable,
+    tables::cmap::{Cmap, CmapSubtable},
+    types::GlyphId,
+};
+
+/// Metadata for constructing a character map from font data.
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub struct CharmapIndex {
+    subtable_offset: u32,
+    is_symbol: bool,
+    is_mac_roman: bool,
+}
+
+impl CharmapIndex {
+    pub(crate) fn new(font: &FontRef<'_>) -> Option<Self> {
+        let cmap = font.cmap().ok()?;
+        let cmap_offset = font
+            .table_directory()
+            .table_records()
+            .iter()
+            .find(|rec| rec.tag() == Cmap::TAG)
+            .map(|rec| rec.offset())?;
+        let (_, rec, _) = cmap.best_subtable()?;
+        let subtable_offset = cmap_offset.checked_add(rec.subtable_offset().to_u32())?;
+        Some(Self {
+            subtable_offset,
+            is_symbol: rec.is_symbol(),
+            is_mac_roman: rec.is_mac_roman(),
+        })
+    }
+
+    /// Creates a character map from the given font data.
+    pub fn charmap<'a>(&self, font_data: &'a [u8]) -> Option<Charmap<'a>> {
+        let subtable_data = font_data.get(self.subtable_offset as usize..)?;
+        let subtable = CmapSubtable::read(FontData::new(subtable_data)).ok()?;
+        Some(Charmap {
+            subtable,
+            is_symbol: self.is_symbol,
+            is_mac_roman: self.is_mac_roman,
+        })
+    }
+}
+
+/// Mapping from Unicode codepoints to nominal glyph identifiers.
+#[derive(Clone)]
+pub struct Charmap<'a> {
+    subtable: CmapSubtable<'a>,
+    is_symbol: bool,
+    is_mac_roman: bool,
+}
+
+impl Charmap<'_> {
+    /// Returns the glyph identifier for the given codepoint.
+    pub fn map(&self, codepoint: impl Into<u32>) -> Option<u32> {
+        let mut c = codepoint.into();
+        if self.is_mac_roman && c > 0x7F {
+            c = unicode_to_mac_roman(c);
+        }
+        let result = match &self.subtable {
+            CmapSubtable::Format0(table) => table.map_codepoint(c),
+            CmapSubtable::Format6(table) => table.map_codepoint(c),
+            CmapSubtable::Format10(table) => {
+                if let Some(index) = c.checked_sub(table.start_char_code()) {
+                    if index < table.num_chars() {
```

Comment:
For my understanding, is this an intentional fast path? As in, it's cheaper to `.num_chars()` than it is to `.glyph_id_array()`?

Reply:
Ah, good catch. This code is copied from harfrust and is something I’d like to move into read-fonts eventually. The extra check here is vestigial from a time when our cmap10 codegen was incorrect (and for some reason, it’s also missing a map_codepoint method).

I’ll fix this before merging and add a todo to get this whole chunk of code into read-fonts so it can be shared my skrifa, harfrust, fontique and others.